KAFKA-4840 : BufferPool errors can cause buffer pool to go into a bad state

smccauliff opened the pull request.

smccauliff committed:
Remove more memory condition variable on throwable or sensor exception.

smccauliff committed:
Test sensor report exception.

smccauliff committed:
Fix checkstyle violations.
Readability.

smccauliff committed:
Restore avaialble memory count on exception.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java:
```diff
@@ -43,6 +43,8 @@
  */
 public class BufferPool {
 
+    static final String BUFFERPOOL_WAIT_TIME_SENSOR_NAME = "bufferpool-wait-time";
```
We don't need the BUFFERPOOL prefix here since it's already defined in the `BufferPool` class. Thanks for the PR, will hopefully review it soon.

smccauliff replied:
Thanks for the feedback.  I've removed this prefix.

smccauliff committed:
Next waiter can check the new available memory, we don't need to have something in the free list.

smccauliff committed:
Only unlock and signal waiters in one place.
Always signal waiters when allocation is complete.
Remove BUFFERPOOL prefix from constant.

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java:
```diff
@@ -43,6 +43,8 @@
  */
 public class BufferPool {
 
+    static final String WAIT_TIME_SENSOR_NAME = "bufferpool-wait-time";
```
I am not sure it is useful to have a static variable for this string since it is used only once. It is also different from the code style in `Sender.SenderMetrics` which assigns sensor name directly. Maybe we can preserve the existing style for simplicity and code style consistence.

smccauliff replied:
There is a mock that refers to this variable otherwise I would not have bothered to put this into a static final variable.

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java:
```diff
@@ -110,78 +112,107 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
+                }
+            }
+        } finally {
+            if (lock.isHeldByCurrentThread()) {
```
Since we only release lock in the `finally` here, do we still need to check `lock.isHeldByCurrentThread`?

smccauliff replied:
Probably not.  The original code was unlocking in different places so this was needed.  I will remove this.

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java:
```diff
@@ -110,78 +112,107 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
-                        timeNs = Math.max(0L, endWaitNs - startWaitNs);
-                        this.waitTime.record(timeNs, time.milliseconds());
-                    }
+                boolean restoreAvailableMemoryOnFailure = true;
```
The name `restoreAvailableMemoryOnFailure` is a bit weird because we should always restore available memory on failure. Maybe we can name it `hasError` and set it to `false` right before `return buffer`.

smccauliff replied:
Sounds good.

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java:
```diff
@@ -110,78 +112,107 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
+                            // just grab a buffer from the free list
+                            buffer = this.free.pollFirst();
+                            accumulated = size;
+                            restoreAvailableMemoryOnFailure = false;
```
We can probably remove this line `restoreAvailableMemoryOnFailure = false`.

smccauliff replied:
Yes.

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java:
```diff
@@ -110,78 +112,107 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
-                    throw new IllegalStateException("Wrong condition: this shouldn't happen.");
+                    // remove the condition for this thread to let the next thread
+                    // in line start getting memory
+                    Condition removed = this.waiters.removeFirst();
```
Should we also remove this condition variable within a `finally` block?

smccauliff replied:
Yes.

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java:
```diff
@@ -110,78 +112,107 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
+                    restoreAvailableMemoryOnFailure = false;
+                    //unlock happens in top-level, enclosing finally
+                    return buffer;
+                } finally {
```
Would it be possible and simpler to merge the two `finally` into one?

smccauliff replied:
Not without additional complications.  One finally deals with cases caused by waiting for memory to become available the other deals with cases causes by just allocating memory.

lindong28 replied:
I think there is actually a reasonable way to merge the two `finally` into one. The idea is that the code in the first finally block doesn't throw exception and is no-op if `this.availableMemory + freeListSize >= size`. This is no-op if you change it to the following:

```
if (hasError)
  this.availableMemory += accumulated;
if (moreMemory != null)
  this.waiters.remove(moreMemory);
```

smccauliff committed:
Consolidate the cleanup of memory allocation into fewer lines of code.

smccauliff committed:
Rename variable.
Don't check to see if lock is held before unlocking, since there is only one place where the lock is unlocked.

smccauliff commented:
@lindong28 Can you check this again?  Thanks.

lindong28 commented:
@smccauliff I think we can further simplify the code by merging the two `finally` into one. But it is OK to me if you don't want to do that. LGTM. Ping @becketqin for review.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java:
```diff
@@ -110,77 +113,70 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                 freeUp(size);
                 this.availableMemory -= size;
-                lock.unlock();
                 return allocateByteBuffer(size);
```
If an OOM is thrown from here, it seems the available memory would be already decremented and not added back.

smccauliff committed:
OOM can cause buffer pool to loose byte when it has memory count available.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java:
```diff
@@ -109,78 +112,72 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
+            // over for them
+            if (!(this.availableMemory == 0 && this.free.isEmpty()) && !this.waiters.isEmpty())
+                this.waiters.peekFirst().signal();
+            lock.unlock();
```
It seems findBugs is not happy with this finally block although I think it should be fine. Can we have a try finally block here to ensure lock.unlock() is executed so that findBugs is happy?

becketqin commented:
@smccauliff Thanks for the update. LGTM. findBugs was complaining about one of the finally block. Can we change it to make findBugs happy?

smccauliff committed:
Another finally otherwise findbugs complains.

becketqin commented:
Thanks for updating the patch. Merged to trunk.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java:
```diff
@@ -109,78 +112,76 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                 // we have enough unallocated or pooled memory to immediately
                 // satisfy the request
                 freeUp(size);
+                ByteBuffer allocatedBuffer = allocateByteBuffer(size);
```
I think it was intentional that we didn't allocate memory with the lock held. It seems like this optimisation was lost?

smccauliff replied:
I'll see if there is some clean way to restore that optimization, but if there is OOM on buffer allocation then the lock needs to be acquired and the allocation freed once more.  Seems pretty ugly.

ijuma replied:
Yeah, it's not pretty. However, reducing the concurrency of `BufferPool` in the common path is not desireable. We definitely need to handle OOMs correctly, but they are relatively rare and it's OK if that path is slower.

becketqin replied:
Good point. I think it is still worth keeping the optimization, although typically the the producer will only allocate poolable batch size, so the actual memory allocation should not happen very often.
